File: vad/open_sat/viterbi_path_mod.py
import numpy as np 

def viterbi_alg(prior, transmat, obslik):
    #return the most-probable (Viterbi) path through 
    #the HMM state trellis
    T = obslik.shape[1]
    Q = prior.shape[0]
    delta, psi, path, scale = np.zeros((Q,T)), np.zeros((Q,T)), np.zeros((1,T)), np.zeros((1,T))
    
    t = 0
    #prior = np.transpose(prior)
    delta[:2,:2] = prior + obslik[:,t]
    psi[:,t] = 0
    for t in np.arange(1,T):
        for j in np.arange(0,Q):
            psi[j,t] = np.argmax(delta[:,t-1] + transmat[:,j])
            delta[j,t] = np.max(delta[:,t-1] + transmat[:,j])
            delta[j,t] += obslik[j,t]
    path[0,T-1] = np.argmax(delta[:,T-1])
    
    for t in np.arange(T-2,-1,-1):
        path[0,t] = psi[int(path[0,t+1]),t+1]
        #print(path[0,t])
    return np.transpose(path)

#if __name__ == '__main__':

File: vad/open_sat/viterbi_path_mod.m
function path = viterbi_path_mod(prior, transmat, obslik)
% VITERBI Find the most-probable (Viterbi) path through the HMM state trellis.
% path = viterbi(prior, transmat, obslik)
%
% Inputs:
% prior(i) = Pr(Q(1) = i)
% transmat(i,j) = Pr(Q(t+1)=j | Q(t)=i)
% obslik(i,t) = Pr(y(t) | Q(t)=i)
%
% Outputs:
% path(t) = q(t), where q1 ... qT is the argmax of the above expression.


% delta(j,t) = prob. of the best sequence of length t-1 and then going to state j, and O(1:t)
% psi(j,t) = the best predecessor state, given that we ended up in state j at t

scaled = 0;

T = size(obslik, 2);
prior = prior(:);
Q = length(prior);

delta = zeros(Q,T);
psi = zeros(Q,T);
path = zeros(1,T);
scale = ones(1,T);


t=1;
delta(:,t) = prior + obslik(:,t);
psi(:,t) = 0; % arbitrary value, since there is no predecessor to t=1
for t=2:T
  for j=1:Q
    [delta(j,t), psi(j,t)] = max(delta(:,t-1) + transmat(:,j));
    delta(j,t) = delta(j,t) + obslik(j,t);
  end
end
[p, path(T)] = max(delta(:,T));
for t=T-1:-1:1
  path(t) = psi(path(t+1),t+1);
end


